types/gasmeter.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2022 Evmos Foundation
+// This file is part of the Evmos Network packages.
+//
+// Evmos is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The Evmos packages are distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the Evmos packages. If not, see https://github.com/evmos/evmos/blob/main/LICENSE
+package types
+
+import (
+	fmt "fmt"
+	math "math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// ErrorNegativeGasConsumed defines an error thrown when the amount of gas refunded results in a
+// negative gas consumed amount.
+// Copied from cosmos-sdk
+type ErrorNegativeGasConsumed struct {
+	Descriptor string
+}
+
+// ErrorGasOverflow defines an error thrown when an action results gas consumption
+// unsigned integer overflow.
+type ErrorGasOverflow struct {
+	Descriptor string
+}
+
+type infiniteGasMeterWithLimit struct {
+	consumed sdk.Gas
+	limit    sdk.Gas
+}
+
+// NewInfiniteGasMeterWithLimit returns a reference to a new infiniteGasMeter.
+func NewInfiniteGasMeterWithLimit(limit sdk.Gas) sdk.GasMeter {
+	return &infiniteGasMeterWithLimit{
+		consumed: 0,
+		limit:    limit,
+	}
+}
+
+// GasConsumed returns the gas consumed from the GasMeter.
+func (g *infiniteGasMeterWithLimit) GasConsumed() sdk.Gas {
+	return g.consumed
+}
+
+// GasConsumedToLimit returns the gas limit if gas consumed is past the limit,
+// otherwise it returns the consumed gas.
+// NOTE: This behavior is only called when recovering from panic when
+// BlockGasMeter consumes gas past the limit.
+func (g *infiniteGasMeterWithLimit) GasConsumedToLimit() sdk.Gas {
+	return g.consumed
+}
```

Comment:
This is the same as `GasConsumed`. @fedekunze Should we refactor it to this or leave it as is:
```suggestion
// GasConsumedToLimit returns the gas limit if gas consumed is past the limit,
// otherwise it returns the consumed gas.
// NOTE: This behavior is only called when recovering from panic when
// BlockGasMeter consumes gas past the limit.
func (g *infiniteGasMeterWithLimit) GasConsumedToLimit() sdk.Gas {
	if g.IsPastLimit() {
		return g.limit
	}
	return g.consumed
}
```

Reply:
need to leave it as is